src/enzyme_ad/jax/Passes/AffineRaising.cpp

```diff
@@ -0,0 +1,462 @@
+//===- AffineRaising.cpp - Raise affine loops to stablehlo ----------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a pass to raise affine loops functions to stablehlo.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/enzyme_ad/jax/Passes/Passes.h"
+
+#include "mlir/Dialect/Affine/Analysis/AffineAnalysis.h"
+#include "mlir/Dialect/Affine/Analysis/Utils.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/IR/AffineValueMap.h"
+#include "mlir/Dialect/Affine/LoopUtils.h"
+#include "mlir/Dialect/Affine/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/IRMapping.h"
+
+#include "src/enzyme_ad/jax/Dialect/Ops.h"
+#include "stablehlo/dialect/StablehloOps.h"
+
+namespace mlir {
+namespace enzyme {
+#define GEN_PASS_DEF_AFFINERAISE
+#include "src/enzyme_ad/jax/Passes/Passes.h.inc"
+} // namespace enzyme
+} // namespace mlir
+
+using namespace mlir;
+using namespace mlir::enzyme;
+
+// Given an affine map for a load/store operation, compute the startIndices,
+// limitIndices and strides corresponding in the memref based on the loop
+// induction variables. (i) -> (0, i, 10) will give [0:1:1, 0:end:1, 10:11:1]
+static LogicalResult affineMapToSlice(affine::AffineValueMap accessValueMap,
+                                      SmallVectorImpl<int64_t> &startIndices,
+                                      SmallVectorImpl<int64_t> &limitIndices,
+                                      SmallVectorImpl<int64_t> &strides) {
+  auto rank = accessValueMap.getNumResults();
+
+  startIndices.reserve(rank);
+  limitIndices.reserve(rank);
+  strides.reserve(rank);
+
+  for (unsigned i = 0; i < rank; i++) {
+    auto expr = accessValueMap.getResult(i);
+    if (auto constExpr = dyn_cast<AffineConstantExpr>(expr)) {
```

Comment:
this will fail for some more complex things like i+1. A couple of suggestions for how to get around:
For the parallel for loop of interest, you can generate a new expression replacing all dims with something else via https://mlir.llvm.org/doxygen/classmlir_1_1AffineExpr.html#abd8b0817ead1cbe3e6bdb194a292272e

For example, you can set startindex = expr.replaceAllDims(dims -> affineParallel.start indices())

and then if that is a constant, you're good! (since presumably expression replace will do the constant folding within the affine expr for you, which is good).

the other ones may be a bit more tricky though